Tests/NIOSOCKSTests/SocksClientHandler+Tests.swift

```diff
@@ -230,5 +236,110 @@ class SocksClientHandlerTests: XCTestCase {
         XCTAssertNoThrow(self.channel.pipeline.addHandler(handler))
         self.assertOutputBuffer([0x05, 0x01, 0x00])
     }
+    
+    func testHandlerRemovalAfterEstablishEvent() {
+        class SOCKSEventHandler: ChannelInboundHandler {
+            typealias InboundIn = NIOAny
+            
+            var establishedPromise: EventLoopPromise<Void>
+            
+            init(establishedPromise: EventLoopPromise<Void>) {
+                self.establishedPromise = establishedPromise
+            }
+            
+            func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+                switch event {
+                case is SOCKSProxyEstablishedEvent:
+                    self.establishedPromise.succeed(())
+                default:
+                    break
+                }
+                context.fireUserInboundEventTriggered(event)
+            }
+        }
+        
+        let establishPromise = self.channel.eventLoop.makePromise(of: Void.self)
+        let removalPromise = self.channel.eventLoop.makePromise(of: Void.self)
+        establishPromise.futureResult.whenSuccess { _ in
+            self.channel.pipeline.removeHandler(self.handler).cascade(to: removalPromise)
+        }
+        
+        try! self.channel.pipeline.addHandler(SOCKSEventHandler(establishedPromise: establishPromise)).wait()
+        
+        self.connect()
+        
+        self.channel.write(ByteBuffer(bytes: [1, 2, 3]), promise: nil)
```

Comment:
Could you just add a comment saying that these writes should be correctly buffered. Also to simplify the test you can probably reduce the amount of buffered data and the flush. This is tested elsewhere.